samples/msal-node-samples/nextjs-rsc-ts/src/utils/AuthProvider.ts

```diff
@@ -0,0 +1,285 @@
+   * @remarks Can safely be called in multiple server components.
+   */
+  authenticate = cache(async () => {
+    const partitionManager = await this.partitionManagerFactory();
+    const homeAccountId = await partitionManager.getKey();
+
+    const instance = await this.getInstance();
+
+    const account = homeAccountId
+      ? await instance.getTokenCache().getAccountByHomeId(homeAccountId)
+      : null;
+
+    return { account, instance };
+  });
+
+  /**
+   * Get the current logged in account.
+   * @returns An account object if a user is logged in, or null if no user is logged in.
+   * @remarks Can safely be called in multiple server components.
+   * @remarks Prefer authenticate() in Server Actions and Route Handlers
+   */
+  getAccount = cache(async () => {
+    const partitionManager = await this.partitionManagerFactory();
+    const homeAccountId = await partitionManager.getKey();
+
+    if (!homeAccountId) {
+      return null;
+    }
+
+    const instance = await this.getInstance();
+
+    return await instance.getTokenCache().getAccountByHomeId(homeAccountId);
+  });
+
+  /**
+   * Get an instance configured with a partitioned cache to the current logged in user.
+   * @returns A ConfidentialClientApplication instance
+   * @remarks Can safely be called in multiple server components.
+   * @remarks Prefer authenticate() in Server Actions and Route Handlers
+   */
+  getInstance = cache(async () => {
+    const cachePlugin = new DistributedCachePlugin(
+      this.cacheClient,
+      await this.partitionManagerFactory()
+    );
+
+    const config = {
+      ...this.configuration,
+      cache: {
+        cachePlugin,
+      },
+    };
+
+    if (!config.auth.cloudDiscoveryMetadata || !config.auth.authorityMetadata) {
+      const metadata = await this.getMetadata(
+        config.auth.clientId,
+        config.auth.authority ?? "https://login.microsoftonline.com/common"
+      );
+
+      if (metadata) {
+        config.auth.cloudDiscoveryMetadata = metadata.cloudDiscoveryMetadata;
+        config.auth.authorityMetadata = metadata.authorityMetadata;
+      }
+    }
+
+    return new ConfidentialClientApplication(config);
+  });
+
+  // validate that the payload includes required fields
+  private getAuthorizationCodePayload(formData: FormData) {
+    // validate that we only get string entries
+    const stringEntries = Array.from(formData.entries()).filter(
+      ([, value]) => typeof value === "string"
+    );
+
+    const data = Object.fromEntries(stringEntries);
+
+    if (!("state" in data)) {
+      throw new Error("No state found in payload.");
+    }
+
+    if (!("code" in data)) {
+      throw new Error("No code found in payload.");
+    }
+
+    return data as Omit<AuthorizationCodePayload, "state"> &
+      Required<Pick<AuthorizationCodePayload, "state">>;
+  }
+
+  private async getMetadata(clientId: string, authority: string) {
+    const tenantId = authority!.split("/").pop()!;
+
+    try {
```

Comment:
```suggestion
            /**
             * If the current msal configuration does not have cloudDiscoveryMetadata or authorityMetadata, we will 
             * make a request to the relevant endpoints to retrieve the metadata. This allows MSAL to avoid making 
             * metadata discovery calls, thereby improving performance of token acquisition process. For more, see:
             * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/performance.md
             */
    try {
```

Reply:
or something similar to the effect

Reply:
Added note where the metadata injection happens, and jsdoc to the metadata methods